src/lib.rs

```diff
@@ -172,81 +172,63 @@
+    #[error(display = "Invalid service start type: {}", _0)]
+    InvalidServiceStartType(u32),
+    /// Invalid raw representation of [`ServiceErrorControl`].
+    #[error(display = "Invalid service error control type: {}", _0)]
+    InvalidServiceErrorControl(u32),
+    /// IO error
+    #[error(display = "IO Error: {}", _0)]
+    IOError(String),
+}
 
-
-pub use error_chain::ChainedError;
-
-error_chain! {
-    errors {
-        /// Invalid account name.
-        InvalidAccountName {
-            description("Invalid account name")
-        }
-        /// Invalid account password.
-        InvalidAccountPassword {
-            description("Invalid account password")
-        }
-        /// Invalid display name.
-        InvalidDisplayName {
-            description("Invalid display name")
-        }
-        /// Invalid database name.
-        InvalidDatabaseName {
-            description("Invalid database name")
-        }
-        /// Invalid executable path.
-        InvalidExecutablePath {
-            description("Invalid executable path")
-        }
-        /// Invalid launch arguments.
-        InvalidLaunchArgument {
-            description("Invalid launch argument")
-        }
-        /// Invalid dependency name.
-        InvalidDependency {
-            description("Invalid dependency name")
-        }
-        /// Invalid machine name.
-        InvalidMachineName {
-            description("Invalid machine name")
-        }
-        /// Invalid service name.
-        InvalidServiceName {
-            description("Invalid service name")
-        }
-        /// Invalid start argument.
-        InvalidStartArgument {
-            description("Invalid start argument")
-        }
-        /// Invalid raw representation of [`ServiceState`].
-        InvalidServiceState(raw_value: u32) {
-            description("Invalid service state")
-            display("Invalid service state value: {}", raw_value)
-        }
-        /// Invalid raw representation of [`ServiceControl`].
-        InvalidServiceControl(raw_value: u32) {
-            description("Invalid service control")
-            display("Invalid service control value: {}", raw_value)
-        }
-        /// Invalid raw representation of [`ServiceStartType`].
-        InvalidServiceStartType(raw_value: u32) {
-            description("Invalid service start type")
-            display("Invalid service start type: {}", raw_value)
-        }
-        /// Invalid raw representation of [`ServiceErrorControl`].
-        InvalidServiceErrorControl(raw_value: u32) {
-            description("Invalid service error control type")
-            display("Invalid service error control type: {}", raw_value)
-        }
-    }
-    foreign_links {
-        System(::std::io::Error) #[doc = "System call error"];
+impl From<std::io::Error> for ErrorKind {
+    fn from(error: std::io::Error) -> Self {
+        ErrorKind::IOError(error.to_string())
```

Comment:
It's probably more appropriate to use the chaining functionality of `Error::source`. This can easily be done with `#[error(cause)]`. See https://github.com/mullvad/hyper-dnscache/blob/master/src/cache_storer.rs#L135 and https://github.com/mullvad/hyper-dnscache/blob/master/src/cache_storer.rs#L31

According to naming convention it should be spelled as `IoError`. However, depending on how it's used, it would probably be nicer for the user if it's split up into multiple errors, and not all errors that happens to originate from an `io::Error` lumped into one. Again, see the links above and compare `WriteFileError` with `ReadFileError`. This allows the error on this level to have a message telling what went wrong on this abstraction, and then the `source` of that error can tell the user what `io::Error` caused that error.

Reply:
i went ahead and broke them all out, i feel like the error names i chose may be a little cumbersome so welcoming any feedback on that as well.